### PR TITLE
Remove old kronmult library. Fall back to CPU implementation.

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This removes the older, slower kronmult library. In the (rare?) case that a GPU kernel is not available it falls back to a CPU kernel.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe): remove rarely used library  

### Does this introduce a breaking change?

- [x] Yes
- [ ] No

In rare cases an input that ran on the GPU may now fall back to the CPU.

## What systems has this change been tested on?

my laptop, Ubuntu 20.04

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/device/kronmult_cuda.cpp [hunk 1]
```diff
@@ -303,101 +294,24 @@ void prepare_kronmult(int const *const flattened_table,
 // note  the input memory referenced by x_ptrs will be over-written
 // --------------------------------------------
 template<typename P>
-void call_kronmult(int const n, P *x_ptrs[], P *output_ptrs[], P *work_ptrs[],
+void call_kronmult(int const n, P *x_ptrs[], P *output_ptrs[],
                    P const *const operator_ptrs[], int const lda,
                    int const num_krons, int const num_dims)
 {
 #ifdef ASGARD_USE_CUDA
+  if (kronmult::is_implemented::gpu(num_dims, n))
```
Comment (mkstoyanov) [inline src/device/kronmult_cuda.cpp]: I was using the old implementation as a fallback for unimplemented `n` I think we should just remove the `is_implemented` struct in `asgard_kronmult.hpp` and here we should just call `execute_gpu()`

The `execute_gpu()` will throw and error if called with incorrect  size.

If you want to use the CPU as a fallback you run into the problem that the data is already loaded on the GPU and will have to be moved back. I think it's OK to throw and in this PR or another we can have a runtime option (command line) to skip the GPU.

What do you think?

**Comment by mkstoyanov** [comment]
Somewhere in ASGarD we are also reporting `workspace` size. We should remove that one too, as it is confusing. (I didn't see it removed in the PR)

**Comment by quantumsteve** [comment]
> Somewhere in ASGarD we are also reporting `workspace` size. We should remove that one too, as it is confusing. (I didn't see it removed in the PR)

I removed the factor of 2. The `element_x` array still uses the workspace size. If we can remove that array, it should be in a separate PR.

File: src/device/kronmult_cuda.cpp (hunk 1 above)
Comment (quantumsteve) [reply src/device/kronmult_cuda.cpp]: :facepalm: The code now checks and throws and exception.

File: src/device/kronmult_cuda.cpp [hunk 2]
```diff
@@ -303,102 +294,25 @@ void prepare_kronmult(int const *const flattened_table,
...
-
-    // -------------------------------------------
-    // note important to wait for kernel to finish
-    // -------------------------------------------
-    auto const stat = cudaDeviceSynchronize();
-    expect(stat == cudaSuccess);
-  }
+  // if a GPU kernel is not available, thow an error
+  if (!kronmult::is_implemented::gpu(num_dims, n))
+    throw std::runtime_error("Kronmult GPU kernel not implemented for degree " +
```
Comment (mkstoyanov) [inline src/device/kronmult_cuda.cpp]: I wasn't clear before, `execute_gpu()` already does the throw. This block should have just the call to execute and the device synch.

The only reason why `is_implemented` was there was to send the unimplemented cases to the old kronmult so the tests wouldn't be failing while I was building the cases.

File: src/kronmult_tests.cpp [hunk 3]
```diff
@@ -119,7 +119,7 @@ TEMPLATE_TEST_CASE("test kronmult", "[kronmult]", float, double)
   SECTION("6d - non-uniform level")
   {
     auto const pde_choice        = PDE_opts::continuity_6;
-    auto const degree            = 1;
+    auto const degree            = 2;
```
Comment (mkstoyanov) [inline src/kronmult_tests.cpp]: Why the change to the test?

File: src/solver_tests.cpp [hunk 4]
```diff
@@ -231,7 +231,7 @@ TEMPLATE_TEST_CASE("test kronmult", "[kronmult]", float, double)
   SECTION("6d - non-uniform level")
   {
     auto const pde_choice        = PDE_opts::continuity_6;
-    auto const degree            = 1;
+    auto const degree            = 2;
```
Comment (mkstoyanov) [inline src/solver_tests.cpp]: Same here, why?

**Review by mkstoyanov — CHANGES_REQUESTED** [review]

**Comment by mkstoyanov** [comment]
So, ignoring the non-RAII memory management in `kronmult_tests.cpp`, do you know why they were reading the workspace back to the host?

File: src/kronmult_tests.cpp (hunk 3 above)
Comment (quantumsteve) [reply src/kronmult_tests.cpp]: An exception was being thrown (no n=1 gpu implementation?)

Comment (mkstoyanov) [reply src/kronmult_tests.cpp]: `asgard_kronmult.cpp` has a d=6 n=1 case implemented (line 391) and there is a test in `asgard_kronmult_tests.cpp` on line 143. What's missing is the `asgard_kronmult.hpp` available in the set (number 106).

The whole `is_implemented` was an ugly hack and should go away (see the earlier comment). That should also fix the test.

File: src/device/kronmult_cuda.cpp (hunk 2 above)
Comment (quantumsteve) [reply src/device/kronmult_cuda.cpp]: removed

File: src/kronmult_tests.cpp (hunk 3 above)
Comment (quantumsteve) [reply src/kronmult_tests.cpp]: `is_implemented` has been removed

**Comment by quantumsteve** [comment]
> So, ignoring the non-RAII memory management in `kronmult_tests.cpp`, do you know why they were reading the workspace back to the host?

It looks like someone wanted to use an existing lambda expression instead of writing a CUDA kernel :man_shrugging:

**Review by mkstoyanov — APPROVED** [review]

**Comment by mkstoyanov** [comment]
I want the old kronmult tests gone too, but you can leave the pleasure for me for when I come back.

**Pull request merged**